src/analysis/processing/qgsalgorithmshortestline.cpp

```diff
@@ -0,0 +1,180 @@
+}
+
+bool QgsShortestLineAlgorithm::prepareAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback * )
+{
+  mSource.reset( parameterAsSource( parameters, QStringLiteral( "SOURCE" ), context ) );
+  if ( !mSource )
+    throw QgsProcessingException( invalidSourceError( parameters, QStringLiteral( "SOURCE" ) ) );
+
+  mDestination.reset( parameterAsSource( parameters, QStringLiteral( "DESTINATION" ), context ) );
+  if ( !mSource )
+    throw QgsProcessingException( invalidSourceError( parameters, QStringLiteral( "DESTINATION" ) ) );
+
+  mMethod = parameterAsInt( parameters, QStringLiteral( "METHOD" ), context );
+
+  mKNeighbors = parameterAsInt( parameters, QStringLiteral( "NEIGHBORS" ), context );
+
+  double paramMaxDist = parameterAsDouble( parameters, QStringLiteral( "DISTANCE" ), context ); //defaults to zero if not set
+  if ( paramMaxDist )
+    mMaxDistance = paramMaxDist;
+
+  return true;
+}
+
+QVariantMap QgsShortestLineAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  if ( mKNeighbors > mDestination->featureCount() )
+    mKNeighbors = mDestination->featureCount();
+
+  QgsFields fields = QgsProcessingUtils::combineFields( mSource->fields(), mDestination->fields() );
+  fields.append( QgsField( QStringLiteral( "distance" ), QVariant::Double ) );
+
+  QString dest;
+  std::unique_ptr< QgsFeatureSink > sink( parameterAsSink( parameters, QStringLiteral( "OUTPUT" ), context, dest, fields, QgsWkbTypes::MultiLineString, mSource->sourceCrs() ) );
+  if ( !sink )
+    throw QgsProcessingException( invalidSinkError( parameters, QStringLiteral( "OUTPUT" ) ) );
+
+  QgsFeatureRequest request = QgsFeatureRequest();
+  request.setDestinationCrs( mSource->sourceCrs(), context.transformContext() );
+
+  QgsSpatialIndex idx = QgsSpatialIndex( mDestination->getFeatures( request ), nullptr, QgsSpatialIndex::FlagStoreFeatureGeometries );
+
+  QgsFeatureIterator sourceIterator = mSource->getFeatures();
+  double step = mSource->featureCount() > 0 ? 100.0 / mSource->featureCount() : 1;
+
+  QgsDistanceArea da = QgsDistanceArea();
+  da.setSourceCrs( mSource->sourceCrs(), context.transformContext() );
+  da.setEllipsoid( context.ellipsoid() );
+
+  int i = 0;
+  QgsFeature sourceFeature;
+  while ( sourceIterator.nextFeature( sourceFeature ) )
+  {
+    if ( feedback->isCanceled() )
+      break;
+
+    const QgsGeometry sourceGeom = sourceFeature.geometry();
+    QgsFeatureIds nearestIds = qgis::listToSet( idx.nearestNeighbor( sourceGeom, mKNeighbors, mMaxDistance ) );
+
+    QgsFeatureRequest targetRequest = QgsFeatureRequest();
+    targetRequest.setFilterFids( nearestIds );
+
+    QgsFeatureIterator destinationIterator = mDestination->getFeatures( targetRequest );
+    QgsFeature destinationFeature;
+    while ( destinationIterator.nextFeature( destinationFeature ) )
```

Comment:
If you want to maximise the speed of this algorithm you can avoid these secondary requests by storing the destination feature attributes at the same time as the spatial index is being built. See qgsalgorithmjoinbynearest.cpp:196 for an example of this approach.